From 3f01d61155ba44c580b8e52a442bcab1d37522fa Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Thu, 13 Mar 2008 20:12:35 -0700
Subject: [PATCH] cavium: udelay optimization for 64bit Cavium CPU.

While not strictly required, Cavium has been using an optimization
in which udelay() is replaced with an asm one.

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/delay.h |   35 +++++++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/delay.h b/include/asm-mips/delay.h
index b0bccd2..d0742cc 100644
--- a/include/asm-mips/delay.h
+++ b/include/asm-mips/delay.h
@@ -16,6 +16,7 @@
 
 #include <asm/compiler.h>
 #include <asm/war.h>
+#include <asm/time.h>
 
 static inline void __delay(unsigned long loops)
 {
@@ -48,6 +49,38 @@ static inline void __delay(unsigned long loops)
 		: "0" (loops), "r" (1));
 }
 
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+
+/*
+ * Replacement for what was being done in get_cycles()
+ * which is being phased out.  If the custom udelay()
+ * goes away, then the octeon_get_cycles should just
+ * revert to using read_c0_count()
+ */
+static inline unsigned long octeon_get_cycles(void)
+{
+	unsigned long result;
+	asm volatile ("rdhwr %0,$31\n"
+#ifndef CONFIG_64BIT
+			"sll %0,0\n"
+#endif
+			: "=r" (result));
+	return result;
+}
+
+/* Implementation of udelay, which unlike the default will actually delay
+    the correct amount of time. Among the other problems with the default
+    spin loop implementation, it overflows at 1ms */
+static inline void udelay(unsigned long usecs)
+{
+	unsigned long cycles = octeon_get_cycles();
+	unsigned long end_cycles = cycles + usecs * (mips_hpt_frequency/1000000);
+	do {
+		cycles = octeon_get_cycles();
+	} while (cycles < end_cycles);
+}
+
+#else
 
 /*
  * Division by multiplication: you don't have to worry about
@@ -109,4 +142,6 @@ static inline void __udelay(unsigned long usecs, unsigned long lpj)
 #define MAX_UDELAY_MS	(1000 / HZ)
 #endif
 
+#endif /* CONFIG_CPU_CAVIUM_OCTEON */
+
 #endif /* _ASM_DELAY_H */
-- 
1.5.5.1

